bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/scheduler2/CronAdjuster.java

```diff
@@ -0,0 +1,551 @@
+/**
+ * Copyright (c) 2014-2017 by the respective copyright holders.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.core.internal.scheduler2;
+
+import java.time.DayOfWeek;
+import java.time.temporal.ChronoField;
+import java.time.temporal.ChronoUnit;
+import java.time.temporal.Temporal;
+import java.time.temporal.TemporalAdjuster;
+import java.time.temporal.TemporalField;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * This class creates a Temporal Adjuster that takes a temporal and adjust it to
+ * the next deadline based on a cron specification.
+ *
+ * @See http://www.cronmaker.com/
+ * @See http://www.quartz-scheduler.org/documentation/quartz-1.x/tutorials/
+ *      crontrigger
+ * @author Peter Kriens - initial contribution and API
+ */
+
+class CronAdjuster implements TemporalAdjuster {
+
+    static Pattern WEEKDAY_P = Pattern.compile("(?<day>\\d+|MON|TUE|WED|THU|FRI|SAT|SUN)(#(?<nr>\\d+)|(?<l>L))?",
+            Pattern.CASE_INSENSITIVE);
+
+    /*
+     * A function interface that we use to check a Temporal to see if it matches
+     * a part of the specification. We combine these checkers in and and ors.
+     */
+    interface Checker {
+        boolean matches(Temporal t);
+    }
+
+    Checker ALWAYS_FALSE = (t) -> false;
```

Comment:
Can mor things be private in this class or maybe moved out and can the order be more strict, static fields, fields, constructors, methods. Then somewhere else the inner classes, depending on where you need to use them.